src/sources/util/tcp/request_limiter.rs

```diff
@@ -0,0 +1,158 @@
+use std::cmp::Ordering;
+use std::sync::{Arc, Mutex};
+use tokio::sync::{OwnedSemaphorePermit, Semaphore};
+
+const EWMA_WEIGHT: f32 = 0.1;
+
+const MINIMUM_PERMITS: usize = 2;
+
+pub struct RequestLimiterPermit {
+    semaphore_permit: Option<OwnedSemaphorePermit>,
+    request_limiter_data: Arc<Mutex<RequestLimiterData>>,
```

Comment:
Since we're already using a tokio semaphore here you might consider using Tokio's mutex. It would play a little better with the scheduler, if nothing else.

Reply:
My understanding is that Tokio's mutex is slower, and the only benefit is that it can be locked across `.await` yield points. The Tokio team themselves recommend the `std::sync::Mutex` in this case.
See this comment for more info: https://github.com/tokio-rs/mini-redis/blob/master/src/db.rs#L40-L51

Reply:
Documentation for the performance difference:
>The feature that the async mutex offers over the blocking mutex is the ability to keep it locked across an .await point. This makes the async mutex more expensive than the blocking mutex, so the blocking mutex should be preferred in the cases where it can be used.

From: https://docs.rs/tokio/latest/tokio/sync/struct.Mutex.html